server/pkg/scene/sceneops/plugin_migrator.go

```diff
@@ -228,11 +175,72 @@ func (s *PluginMigrator) MigratePlugins(ctx context.Context, sc *scene.Scene, ol
 		Scene:             sc,
 		Layers:            modifiedLayers,
 		Properties:        properties,
-		RemovedLayers:     removedLayers,
+		RemovedLayers:     removedLayers.Layers(),
 		RemovedProperties: removedPropertyIDs,
 	}, nil
 }
 
+func isInvalidLayer(llp *layer.PluginID, lle *layer.PluginExtensionID, oldPlugin, newPlugin *plugin.Plugin) bool {
+	// a layer considered invalid if it is related to the old plugin
+	// and if the new plugin version does not support the extension
+	return llp != nil && lle != nil && (*llp).Equal(oldPlugin.ID()) && newPlugin.Extension(*lle) == nil
+}
+
+func collectLayerPropertyIds(ll layer.Layer) []property.ID {
+	removedPropertyIDs := []property.ID{}
+	if p := ll.Property(); p != nil {
+		removedPropertyIDs = append(removedPropertyIDs, *p)
+	}
+	if ib := ll.Infobox(); ib != nil {
+		removedPropertyIDs = append(removedPropertyIDs, ib.Property())
+		for _, f := range ib.Fields() {
+			removedPropertyIDs = append(removedPropertyIDs, f.Property())
+		}
+	}
+	return removedPropertyIDs
+}
+
+func (s *PluginMigrator) loadSchemas(ctx context.Context, oldPlugin *plugin.Plugin, newPlugin *plugin.Plugin) (map[property.SchemaID]*property.Schema, error) {
+	schemaIDs := collectSchemaIds(oldPlugin, newPlugin)
+	schemas, err := s.PropertySchema(ctx, schemaIDs...)
+	if err != nil {
+		return nil, err
+	}
+	schemaMap := map[property.SchemaID]*property.Schema{}
+	if opsId := oldPlugin.Schema(); opsId != nil {
+		if npsId := newPlugin.Schema(); npsId != nil {
+			for _, s := range schemas {
+				if s.ID() == *npsId {
+					schemaMap[*opsId] = s
+				}
+			}
+		}
+	}
+	for _, e := range oldPlugin.Extensions() {
+		if npe := newPlugin.Extension(e.ID()); npe != nil {
+			for _, s := range schemas {
+				if s.ID() == npe.Schema() {
+					schemaMap[e.Schema()] = s
+				}
+			}
+		}
+	}
+	return schemaMap, nil
+}
+
+func collectSchemaIds(oldPlugin *plugin.Plugin, newPlugin *plugin.Plugin) []property.SchemaID {
+	schemaIDs := []property.SchemaID{}
+	if oldPlugin.Schema() != nil {
+		if npsId := newPlugin.Schema(); npsId != nil {
+			schemaIDs = append(schemaIDs, *npsId)
+		}
+	}
+	for _, e := range newPlugin.Extensions() {
+		schemaIDs = append(schemaIDs, e.Schema())
+	}
+	return schemaIDs
+}
```

Comment:
`collectSchemaIds` is no longer needed. Use `PropertySchemas` method in `plugin.Plugin`. Also using `property.SchemaIDSet` is helpful.

```go
schemaSet := property.NewSchemaIDSet(oldPlugin.PropertySchemas()...)
schemaSet.Add(newPlugin.PropertySchemas()...)
schemas := schemaSet.List()
```

And, you can use `Set` method in `idx.List` by using the latest version of `reearthx`

```go
schemas := oldPlugin.PropertySchemas().Set().Concat(newPlugin.PropertySchemas().Set()).List()
```